Collect metrics on time spent processing repeat records (excluding waiting for endpoints)

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-15798

We recently implemented rate limiting on repeaters based on time spent waiting for an endpoint, but are surprised to see slower timings even for rate limited requests. This should give us more insight into the time spent processing repeat records on our side, subtracting the time spent waiting for an endpoint to respond if it is an attempted forward.

If we leave these metrics in place longer term, we likely do not need to collect the current metrics around rate limited repeat records since this also tracks that, but I wanted to leave room to iterate on this before removing any existing metrics.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Track time spent processing repeat records

We rate limit based on time spent waiting for an endpoint to respond,
but would like to have more insight on how much time is being spent
outside of waiting for the endpoint since that is theoretically within
our control.

gherceg committed:
Tag timings with action

dannyroberts commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -143,39 +144,57 @@ def retry_process_repeat_record(repeat_record_id, domain):
+                repeat_record.postpone_by(random.uniform(*RATE_LIMITER_DELAY_RANGE))
+                action = 'rate_limited'
+            elif repeat_record.is_queued():
+                with TimingContext() as fire_timer:
```
any reason not to pass in a name for this timing context?

gherceg replied:
I knew I wouldn't be referencing it by name anywhere so I didn't, but it does only add confusion so might as well add a name.

gherceg replied:
And I recognize I don't reference the `endpoint_timing` context by name either, though I was hopeful that I would use it whether for asserting I have the right context or finding a better way to grab the context in the first place.

dannyroberts commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -143,39 +144,57 @@ def retry_process_repeat_record(repeat_record_id, domain):
+                    repeat_record.fire()
+                # round up to the nearest millisecond, meaning always at least 1ms
+                report_repeater_usage(repeat_record.domain, milliseconds=int(fire_timer.duration * 1000) + 1)
+                time_spent_waiting = fire_timer.subs[0].duration
```
is there a way to refer to the particular sub-timer by name? Hard for me to tell if this refers to the same thing as fire_timer.duration, or refers to the `endpoint_timing` above

gherceg replied:
yeah I didn't love this. In hindsight, I can convert it to a dictionary and navigate the subs field that way.

gherceg replied:
Eh I didn't think that through fully. Ended up using a list comprehension inbaea03a96b2b5882e575ada962b506b6d0b8d291. Still doesn't feel great but short of adding a method to the TimingContext, I think it is the best option.

gherceg committed:
Update units for metrics_historgram

Pass buckets, bucket_tag, and bucket_unit

gherceg committed:
Get endpoint timer by name

gherceg commented:
I do think it is worth creating one more upper bucket (5 seconds or something like that, since we have seen tasks that are greater than a second and less than ten. Would be nice to get more insight there.

gherceg committed:
Change bucket boundaries

I don't think we care (at least for now) about differentiating between
buckets of 10 and 30 ms. I added a 5 second bucket, and removed two of
the lower buckets.

gherceg committed:
Time request in fire_for_record and pass up

I missed that you needed to pass a timing context down the call stack in
order to take advantage of nestable timers. Rather than pass a context
down, lets just time the request in fire_for_record and pass that value
back up.

dannyroberts commented on corehq/motech/repeaters/models.py:
```diff
@@ -434,22 +435,32 @@ def retire(self):
+        # time how long the request takes
+        start_time = time.time()
         try:
             response = self.send_request(repeat_record, payload)
```
Is what you did here different from putting a timing context around just this one line?

gherceg replied:
ugh I had it in my head that the timer wouldn't stop until the exception was handled and I'm not sure why. Tested in a shell and behaves as expected, so made the change 4183245cc9211c1a460addbeffff1921e6ec1012

gherceg committed:
Just use a TimingContext :facepalm:

gherceg commented:
Going to make a substantial change to this so reviewers can hold off for (looking at you @millerdev)

gherceg committed:
Pass timing context down instead of duration up

It is less disruptive to pass an optional timing context down the call
stack since the caller can access the subtimer from the context it owns.
The alternative requires changing the return signature of multiple
functions, impacting callers that don't actually care about timing.

gherceg committed:
Cleanup logic to grab duration from sub timer

gherceg committed:
Merge branch 'master' into gh/repeaters/improve-timing-metrics

gherceg committed:
Merge branch 'master' into gh/repeaters/improve-timing-metrics

gherceg commented:
@millerdev and @dannyroberts this is ready to review now.

millerdev commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -143,39 +145,59 @@ def retry_process_repeat_record(repeat_record_id, domain):
+                # round up to the nearest millisecond, meaning always at least 1ms
+                report_repeater_usage(repeat_record.domain, milliseconds=int(fire_timer.duration * 1000) + 1)
+                action = 'attempted'
+                request_duration = [sub.duration for sub in fire_timer.root.subs if sub.name == ENDPOINT_TIMER][0]
```
Is there any chance that the list comprehension would result in an empty list? I think that would cause an error. Here's a suggestion to avoid that:
```suggestion
                request_duration = sum(sub.duration for sub in fire_timer.root.subs if sub.name == ENDPOINT_TIMER)
```

gherceg replied:
Yeah this was the least favorite part of my changes here, since it feels like there should be a nicer way to get sub timers. Given we pass the timing context into `repeat_record.fire` right above, I felt like it should be safe to assume there will always be a subtimer available (otherwise, if the timing code in `fire` is removed, this should fail loudly so that we remove it here as well? If it handles the lack of a timer gracefully, we might end up collecting data that we think is telling us one thing, but in fact just doesn't include timing the endpoint request at all. What do you think?

millerdev replied:
I see, so this code structure is functioning partly as an assertion. Seems reasonable as long as there is no way for it to fail without a bug in the code.

gherceg committed:
Flatten out list of sub timers

gherceg commented:
Confirmed this is working as expected: https://app.datadoghq.com/dashboard/u88-px4-jra/repeaters-dashboard?fromUser=false&fullscreen_end_ts=1723595561362&fullscreen_paused=false&fullscreen_refresh_mode=sliding&fullscreen_section=overview&fullscreen_start_ts=1723581161362&fullscreen_widget=3480984443272320&refresh_mode=sliding&tpl_var_environment%5B0%5D=staging&view=spans&from_ts=1723581101971&to_ts=1723595501971&live=true
